src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py

```diff
@@ -1556,8 +1558,26 @@ def aks_create(cmd, client, resource_group_name, name, ssh_key_value,  # pylint:
     retry_exception = Exception(None)
     for _ in range(0, max_retry):
         try:
-            return sdk_no_wait(no_wait, client.create_or_update,
-                               resource_group_name=resource_group_name, resource_name=name, parameters=mc)
+            result = sdk_no_wait(no_wait,
+                                 client.create_or_update,
+                                 resource_group_name=resource_group_name,
+                                 resource_name=name, parameters=mc)
+        # add cluster spn with Monitoring Metrics Publisher role assignment to the cluster resource
+        # mdm metrics supported only in azure public cloud so add the  role assignment only in this cloud
+            cloud_name = cmd.cli_ctx.cloud.name
+            if cloud_name.lower() == 'azurecloud' and monitoring:
+                from msrestazure.tools import resource_id
+                cluster_resource_id = resource_id(
+                    subscription=subscription_id,
+                    resource_group=resource_group_name,
+                    namespace='Microsoft.ContainerService', type='managedClusters',
+                    name=name
+                )
+                if not _add_role_assignment(cmd.cli_ctx, 'Monitoring Metrics Publisher',
+                                            service_principal_profile.client_id, scope=cluster_resource_id):
+                    logger.warning('Could not create a role assignment for monitoring addon. '
+                                   'Are you an Owner on this subscription?')
+            return result
```

Comment:
Will this work when `--no-wait` is actually specified? I thought that to ensure the cluster RG is already created, we would need to hook into some type of callback from the `sdk_no_wait` function.

Reply:
Yes,Matt,  I tested with --no-wait and without. Both scenarios working as expected. Since the roll assignment is happening after the invocation of create_or_update API.  I reviewed the python code of create_or_update API,  response returned after the successful call so I think, we are good.   call_or_update API is on the call back of sdk_no_wait.